### PR TITLE
Refactoring invoiceMapperService Test

### PR DESCRIPTION
**Pull request by tegar-wahyu**
### Refactor: Split InvoiceMapperService Tests into Modular Files

Pull request ini bertujuan untuk memisahkan unit test `InvoiceMapperService` yang sebelumnya terkonsentrasi dalam satu file (`invoiceMapperService.test.js`) menjadi beberapa file terpisah dalam folder baru. Perubahan ini meningkatkan organisasi kode, mempermudah perawatan, dan mendukung skalabilitas.

#### Perubahan Utama
1. **Pemisahan File Test**:
   - Semua test untuk `InvoiceMapperService` kini diorganisir dalam folder baru: `backend/tests/services/InvoiceMapperService/`.
   - Setiap file test fokus pada aspek spesifik dari layanan, mengikuti prinsip **Single Responsibility** untuk mempermudah debugging dan perawatan.

2. **Penambahan File Unit Test**:
   - `backend/tests/mocks/models.mock.js`: File mock baru untuk menyediakan data tiruan (mock data) yang digunakan dalam test.
   - `backend/tests/services/InvoiceMapperService/PartnerIdGeneration.test.js`: Menguji logika pembuatan `partnerId` dari nama vendor.
   - `backend/tests/services/InvoiceMapperService/addressProcessing.test.js`: Menguji logika pemrosesan dan pemetaan alamat.
   - `backend/tests/services/InvoiceMapperService/coreMapping.test.js`: Menguji fungsi pemetaan inti dari hasil OCR ke model `Invoice`.
   - `backend/tests/services/InvoiceMapperService/dataProcessing.test.js`: Menguji pemrosesan data umum.
   - `backend/tests/services/InvoiceMapperService/dateFormatParsing.test.js`: Menguji parsing format tanggal.
   - `backend/tests/services/InvoiceMapperService/extractInvoiceNumber.test.js`: Menguji ekstraksi nomor invoice.
   - `backend/tests/services/InvoiceMapperService/fieldParsing.test.js`: Menguji ekstraksi dan pemetaan field generik.
   - `backend/tests/services/InvoiceMapperService/lineItemProcessing.test.js`: Menguji pemrosesan item baris (line items) pada invoice.
   - `backend/tests/services/InvoiceMapperService/inputValidation.test.js`: Menguji validasi input untuk memastikan data yang diterima sesuai.
   - `backend/tests/services/InvoiceMapperService/setupAzureInvoiceMapper.js`: File setup terpusat untuk menginisialisasi `AzureInvoiceMapper` dan konfigurasi Test Double (stub) untuk dependensi eksternal.

3. **Penghapusan File**:
   - `backend/tests/services/invoiceMapperService.test.js`: File ini dihapus karena semua test telah dipindahkan ke file-file baru yang lebih terorganisir. File ini sebelumnya berisi lebih dari 1000 baris kode, yang membuatnya sulit untuk dikelola.

#### Manfaat Perubahan
- **Organisasi yang Lebih Baik**:
  - Test kini dikelompokkan berdasarkan fungsionalitas (misalnya, pemrosesan alamat, parsing tanggal, validasi input), sehingga lebih mudah untuk menemukan dan memperbarui test terkait fitur tertentu.
- **Kemudahan Maintenance**:
  - File yang lebih kecil dan fokus mengurangi kompleksitas, mempermudah debugging, dan mempercepat isolasi masalah ketika test gagal.
- **Skalabilitas**:
  - Struktur modular memungkinkan penambahan test baru tanpa mengganggu file yang sudah ada. Misalnya, menambahkan test untuk format mata uang baru hanya memerlukan perubahan pada `fieldParsing.test.js`.
- **Efisiensi Test**:
  - File yang lebih kecil memungkinkan eksekusi paralel yang lebih efisien dengan Jest, mengurangi waktu test.
- **Kontrol Versi yang Lebih Baik**:
  - File yang terpisah mengurangi risiko merge conflicts saat beberapa orang bekerja pada test yang berbeda.
- **Dokumentasi yang Lebih Jelas**:
  - Struktur file berfungsi sebagai dokumentasi, membantu pembaca memahami kemampuan `InvoiceMapperService` dengan cepat.
 
 ---

Note: `addressProcessing.test.js` belum menggunakan implementasi mapping alamat yang baru sehingga perlu dilakukan penyesuaian

### CODE REVIEW

**Comment by tegar-wahyu** [comment]
Update: `addressProcessing.test.js` sudah menggunakan implementasi mapping alamat yang baru

**Comment by suryata** [comment]
@codecov-ai-reviewer review

File: backend/tests/services/InvoiceMapperService/setupAzureInvoiceMapper.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  return partnerId;
+};
+
+const partnerId = "contoso-partner";
+
+module.exports = {
+  getMapper: () => mapper,
+  partnerId,
+};
```
[the remark this answers is not in the document]
Comment (tegar-wahyu) [reply backend/tests/services/InvoiceMapperService/setupAzureInvoiceMapper.js]: Resolved in: 8f36848: [REFACTOR] Enhance test setup for AzureInvoiceMapper with configurable mapper instance

File: backend/tests/services/InvoiceMapperService/lineItemProcessing.test.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      
+      const nullContentItemsField = {
+        someProperty: 'test',
+        content: null
+      };
+      
+      
+      expect(mapper.extractLineItems(nullContentItemsField)).toEqual([]);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (tegar-wahyu) [reply backend/tests/services/InvoiceMapperService/lineItemProcessing.test.js]: Resolved in: c1775e3: [TEST] Add test case for handling negative quantities and zero prices in line item processing

File: backend/tests/mocks/models.mock.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+jest.mock('../../../src/models', () => ({
+  Invoice: {
+    build: jest.fn().mockImplementation(data => data)
+  },
+  sequelize: {}
+}));
```
[the remark this answers is not in the document]
Comment (tegar-wahyu) [reply backend/tests/mocks/models.mock.js]: Resolved in: ee2b0a4: [REFACTOR] Enhance mock implementation for models with additional methods for testing

File: backend/tests/services/InvoiceMapperService/dateFormatParsing.test.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      jest.spyOn(global, 'Date').mockImplementation(() => now);
+      
+      // Test invalid format that should fall back to current date
+      const invalidDate = mapper.parseDate({ content: 'not-a-date' });
+      expect(invalidDate).toEqual(now);
+      
+      // Clean up
+      jest.restoreAllMocks();
+    });
+  });
```
[the remark this answers is not in the document]
Comment (tegar-wahyu) [reply backend/tests/services/InvoiceMapperService/dateFormatParsing.test.js]: Resolved in : 23ea0bb: [REFACTOR] Update date format parsing test to use fake timers for consistent results

File: backend/tests/services/InvoiceMapperService/coreMapping.test.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      const { invoiceData } = await mapper.processInvoiceData(ocrResult, partnerId, null);
+      expect(invoiceData.file_url).toBeNull();
+      expect(invoiceData.invoice_number).toBe('INV-TEST-NULL-URL');
+      expect(invoiceData.partner_id).toBe(partnerId);
+      
+      // Test omitting fileUrl parameter
+      const { invoiceData: invoiceData2 } = await mapper.processInvoiceData(ocrResult, partnerId);
+      expect(invoiceData2.file_url).toBeNull();
+    });
+  });
```
[the remark this answers is not in the document]
Comment (tegar-wahyu) [reply backend/tests/services/InvoiceMapperService/coreMapping.test.js]: Resolved in: 59d8780: [REFACTOR] Split tests for core mapping functionality and currency handling in InvoiceMapperService

File: backend/tests/services/InvoiceMapperService/addressProcessing.test.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        VendorName: { content: 'Test Vendor' },
+        VendorTaxId: { content: '123-45-6789' }
+      };
+      
+      const result = mapper.extractVendorData(fields);
+      expect(result.address).toBe('123 Business St Chicago, IL 60601');
+      expect(result.name).toBe('Test Vendor');
+      expect(result.tax_id).toBe('123-45-6789');
+    });
+});
```
[the remark this answers is not in the document]
Comment (tegar-wahyu) [reply backend/tests/services/InvoiceMapperService/addressProcessing.test.js]: Resolved in: e50e95f: [REFACTOR] Improve address processing tests with structured test cases and enhanced setup

**Comment by tegar-wahyu** [comment]
> Halo Tegar! 👋
> 
> Terima kasih atas PR refactoring Invoice Service-nya! Test file untuk core mapping ini udah lumayan lengkap, tapi file-nya sepertinya masih cukup besar untuk dihandle (188 baris) dan berisi beberapa aspek sekaligus, mulai dari pengecekan parsing mata uang, validasi field, sampai mapping. Agar maintainabilitynya makin oke dan gampang untuk menemukan test yang spesifik, saya sarankan untuk memecah file ini menjadi beberapa file yang lebih fokus, misalnya:
> 
> 1. currencyHandling.test.js – khusus buat pengujian parsing dan fallback currency,
> 2. fieldParsing.test.js – untuk validasi dan pengolahan field,
> 3. coreMappingBasics.test.js – untuk pengujian dasar pemetaan dari hasil OCR ke model invoice.
> 
> Selain itu, untuk file test date format parsing, ada beberapa masukan dari saya. Implementasi mocking tanggalnya masih bisa di-improve. Daripada langsung memodifikasi objek global Date, coba deh pake Jest fake timers (jest.useFakeTimers()) dan set waktu sistem secara spesifik. Cara ini bakal lebih reliable dan menghindari potensi efek samping di test lain.
> 
> Terima kasih atas kerjanya yang luar biasa, semangat terus refactoring-nya! 🙌

## PR Resolution Response

Thank you reviewnya Surya! Saya sudah mengimplemntasikan semua improvements yang disarankan:

### Test File Refactoring
Memisahkan test file besar (188) baris menjadi 3 file:
- `currencyHandling.test.js` - Currency parsing dan fallback tests
- `fieldParsing.test.js` - Handle validasi field dan processing tests
- `coreMappingBasics.test.js` - Fokus pada inti OCR-to-invoice model mapping tests

Pemisahan ini menjadikan test lebih maintanable dan jadi lebih mudah untuk mencari fungsionalitas tertentu.

### Date Mocking Improvement
Untuk test date format parsing, Saya mengganti direct global Date object modification dengan fake timer nya Jestseperti yang disarankan:

```javascript
beforeEach(() => {
  jest.useFakeTimers();
  jest.setSystemTime(new Date('2023-05-15'));
});

afterEach(() => {
  jest.useRealTimers();
});
```

Semua tes sudah pass 100%. Terima kasih feedbacknya Surya!🙏

**Comment by asteriskzie** [comment]
Halo Tegar, setelah merge dev dengan PR #69 ternyata ada conflict yang perlu diselesaikan dulu
